### PR TITLE
Refactored gaussian_euclidean and added some tests

### PR DESCRIPTION
**Pull request by canyon289**
Getting some mypy failures that I think are unrelated to the code change but can investigate later

### CODE REVIEW

**Comment by canyon289** [comment]
Huh interesting I get errors locally but seems like were good in CI

image

File: tests/test_metrics.py
Comment (MarcoGorelli) [inline tests/test_metrics.py:13]: this is (`monkeypatch`) used here?

Comment (canyon289) [reply tests/test_metrics.py:13]: Yes its monkeypatch? I feel like I'm missing a question hee

Comment (MarcoGorelli) [reply tests/test_metrics.py:13]: sorry, ignore me, you have `monkeypatch.setattr(metrics.jnp, "multiply", multiply)`, not sure what I was thinking :flushed:

Comment (canyon289) [reply tests/test_metrics.py:13]: No worries! I appreciate the review :)

File: blackjax/inference/metrics.py
Comment (rlouf) [inline blackjax/inference/metrics.py:66]: Nice, I like this! You should raise an error if `ndim=0` as well. The following flow would work:

```python
if ndim == 1:
    dot, matmul = jnp.multiply, jnp.multiply
elif ndim==2:
    dot, matmul = jnp.dot, jnp.matmul
else:
    raise ValueError
    
# Everything else
```

File: tests/test_metrics.py
Comment (rlouf) [inline tests/test_metrics.py:36]: I learned a lot reading this!

I am wondering if instead of monkey patching the methods we could implement some functional tests? It would be simpler and test the whole chain. The issue with very granular tests is that you end up fighting against them every time you make a small change that doesn't affect the rest of the library or the results.

I tend to only test interfaces; tests are here to guarantee a contract that you make with users and other developers that are interfacing with this code. This goes both ways: the output of your function should be "correct", that includes making sure at initialization that you are not going to make computations with garbage. How about making sure here that:

1. An exception is raised with the wrong number of dimensions;
2. For 1- and 2-dimensional arrays we get values for the momentum and kinetic energy that are correct;

**Review by rlouf — CHANGES_REQUESTED** [review]
Sorry for coming back to this so late! I like that you simplified, thus improved the `metrics` code. I have some remarks about the tests; what you did with monkey patching is impressive but I think you may be missing the target.

File: tests/test_metrics.py
Comment (canyon289) [reply tests/test_metrics.py:36]: First off happy to change it and will do so sometime soon. Anything I say after this is just for conversation.

Dont interface tests suffer from the condition youre mentioning about fighting the testing suite after a small change? If someone makes a chance to a small piece one line of a function now you have to dig through a bunch of integration tests to figure out where it broke, versus with smaller unit tests only the the piece that broke will have an issue?

This is not an argument against integration tests, theyre great for all the reasons you mention. In my experience having both has been helpful so even if two things fail, at least one of them tells me specifically what is failing that helps fix everything else.

File: tests/test_metrics.py
Comment (canyon289) [inline tests/test_metrics.py:60]: @rlouf This one better? If so I'll test kinetic energy, add an ndim 2 test and make all the other changes too

File: blackjax/inference/metrics.py
Comment (canyon289) [inline blackjax/inference/metrics.py:87]: What does std stand for? Standard deviation?

File: blackjax/inference/metrics.py
Comment (canyon289) [reply blackjax/inference/metrics.py:66]: Will do

File: blackjax/inference/metrics.py
Comment (rlouf) [reply blackjax/inference/metrics.py:87]: Yeah. It's a terrible name when you think too much about it.

Comment (canyon289) [reply blackjax/inference/metrics.py:87]: Maybe it should be named something like random sample, or momentum_sample?

Comment (rlouf) [reply blackjax/inference/metrics.py:87]: Oh no it stands for standard normal I believe.

File: tests/test_metrics.py
Comment (rlouf) [reply tests/test_metrics.py:36]: If your interfaces are well-designed they don't; with a mature library interfaces are less likely to change than what they enclose. I would argue you *want* the tests to break when there is a change of interface rather, but not when you add a numerical optimization somewhere for instance. Unit tests would break on the latter.

I am not against unit tests in general; I find that they are very useful when you are first building the library as it makes you confident that you are moving in the right direction. But eventually they become a burden. Plus when you make a modification later, you know exactly what broke the tests on the interface: the modification you just made.

Comment (rlouf) [reply tests/test_metrics.py:36]: You want to optimize for developer speed and test suite running speed as much as you can. When  you have too many unit tests they tend to get in the way of developers and slow down the CI.

File: tests/test_metrics.py
Comment (rlouf) [reply tests/test_metrics.py:60]: Yes! You can add a few edge cases that you think are important. For instance, when testing distributions in MCX I made sure that for a random variable with a stricly positive support `logpdf(0)` returned `-inf` as this is an important edge case.

File: blackjax/inference/metrics.py
Comment (canyon289) [reply blackjax/inference/metrics.py:87]: Would you be opposed if it was renamed standard_normal to make it unambiguous?

**Comment by rlouf** [comment]
Explicit is always better :)

File: tests/test_metrics.py
Comment (canyon289) [reply tests/test_metrics.py]: What would the edge cases here be? Negative mass matrix? Other than that the other calculations look pretty straightforward

**Comment by canyon289** [comment]
Can I merge this now?

**Comment by rlouf** [comment]
Will have a look this morning, I'm sorry for no having done this earlier.

File: blackjax/inference/metrics.py
Comment (rlouf) [inline blackjax/inference/metrics.py:71]: remove empty line?

**Comment by rlouf** [comment]
It looks great! Would you mind removing the empty line after `if ndim==2`? You can merge once that's done.

File: blackjax/inference/metrics.py
Comment (canyon289) [reply blackjax/inference/metrics.py]: Done, will merge after test pass

**Pull request merged**
